d3a_api_client/grid_fee_calculation.py

```diff
@@ -0,0 +1,102 @@
+import logging
+from copy import copy
+
+from d3a_interface.utils import key_in_dict_and_not_none
+
+
+class GridFeeCalculation:
+
+    def __init__(self):
+        self.latest_grid_stats_tree = {}
+        self.paths_to_root_mapping = {}
+        self.market_name_grid_fee_mapping = {"last_market_fee": {},
+                                             "next_market_fee": {}}
+
+    def _handle_grid_stats(self, message):
+        for device_event in message["content"]:
+            if device_event["event"] == "grid stats":
+                self.latest_grid_stats_tree = device_event["grid_stats_tree"]
+                message["content"].remove(device_event)
+        self._populate_grid_fee_mappings()
+
+    def _populate_grid_fee_mappings(self):
+        for fee_type in ["last_market_fee", "next_market_fee"]:
+            self._get_grid_fee_area_mapping_from_dict(self.latest_grid_stats_tree, fee_type)
+
+        self._get_all_paths_in_grid_stats_dict(self.latest_grid_stats_tree, [])
```

Comment:
I think this is not optimal, because these methods walk 3 times over the grid tree, while just once would be enough. Why not having just one walk over the tree nodes, and populate the class members while being on every node of the tree? For instance, one method could populate the market_name_grid_fee_mapping variable for both last_market_fee, next_market_fee, and also the self.paths_to_root_mapping in one go.